runners/google-cloud-dataflow-java/src/main/java/org/apache/beam/runners/dataflow/util/PackageUtil.java

```diff
@@ -144,48 +136,73 @@ static PackageAttributes createPackageAttributes(File classpathElement,
-          + "of an issue in your pipeline. You may want to consider trimming the classpath to "
-          + "necessary dependencies only, using --filesToStage pipeline option to override "
-          + "what files are being staged, or bundling several dependencies into one.",
+              + "copies to all workers. Having this many entries on your classpath may be indicative "
+              + "of an issue in your pipeline. You may want to consider trimming the classpath to "
+              + "necessary dependencies only, using --filesToStage pipeline option to override "
+              + "what files are being staged, or bundling several dependencies into one.",
           classpathElements.size());
     }
 
-    ArrayList<DataflowPackage> packages = new ArrayList<>();
-
     if (stagingPath == null) {
       throw new IllegalArgumentException(
           "Can't stage classpath elements on because no staging location has been provided");
     }
 
-    int numUploaded = 0;
-    int numCached = 0;
+    ListeningExecutorService executorService = MoreExecutors.listeningDecorator(
+        Executors.newFixedThreadPool(32));
+    List<ListenableFuture<PackageAttributes>> futures = new LinkedList<>();
     for (String classpathElement : classpathElements) {
-      String packageName = null;
+      String userPackageName = null;
       if (classpathElement.contains("=")) {
         String[] components = classpathElement.split("=", 2);
-        packageName = components[0];
+        userPackageName = components[0];
         classpathElement = components[1];
       }
+      final String packageName = userPackageName;
 
-      File file = new File(classpathElement);
+      final File file = new File(classpathElement);
       if (!file.exists()) {
         LOG.warn("Skipping non-existent classpath element {} that was specified.",
             classpathElement);
         continue;
       }
 
-      PackageAttributes attributes = createPackageAttributes(file, stagingPath, packageName);
+      ListenableFuture<PackageAttributes> future =
+          executorService.submit(new Callable<PackageAttributes>() {
+            @Override
+            public PackageAttributes call() throws Exception {
+              return createPackageAttributes(file, stagingPath, packageName);
+            }
+          });
+      futures.add(future);
+    }
+    List<PackageAttributes> packageAttributes;
+    try {
+      packageAttributes = Futures.allAsList(futures).get();
+    } catch (InterruptedException e) {
+      Thread.currentThread().interrupt();
+      throw new RuntimeException("Interrupted while staging packages", e);
+    } catch (ExecutionException e) {
+      throw new RuntimeException("Error while staging packages", e.getCause());
+    } finally {
+      executorService.shutdown();
+    }
 
+    List<DataflowPackage> packages = new ArrayList<>();
+    int numUploaded = 0;
+    int numCached = 0;
+
+    for (PackageAttributes attributes : packageAttributes){
```

Comment:
The real performance win can be found by parallelizing this loop.

We can get around the upload buffer size limit by doing an instanceof check and down casting to the GcsWriteChannel and setting the upload buffer size to be some sane limit like the smaller of 1mb or 1 byte greater than the file size.

Finally if we sort the file upload order to prioritize the largest files first, we can prevent the tail being dominated by a single large file that can't be split.